web/src/lib/components/memory-page/memory-viewer.svelte

```diff
@@ -42,119 +45,108 @@
 
 		const queryIndex = $page.url.searchParams.get('index');
 		if (queryIndex != null) {
-			currentIndex = parseInt(queryIndex);
-			if (isNaN(currentIndex) || currentIndex > $memoryStore.length - 1) {
-				currentIndex = 0;
+			memoryIndex = parseInt(queryIndex);
+			if (isNaN(memoryIndex) || memoryIndex > $memoryStore.length - 1) {
+				memoryIndex = 0;
 			}
 		}
+	});
 
-		currentMemory = $memoryStore[currentIndex];
+	onDestroy(async () => {
+		if (browser) pause();
+	});
 
-		nextIndex = currentIndex + 1;
-		nextMemory = $memoryStore[nextIndex];
+	const canAdvance = () => !!(nextMemory || nextAsset);
 
-		if (currentIndex > 0) {
-			lastMemory = $memoryStore[lastIndex];
-		}
-	});
+	const duration = 5000; // 5 seconds
 
-	const toNextMemory = (): boolean => {
-		if (showNextMemory) {
-			resetAutoPlay();
+	let paused = true;
+	let progress = 0;
+	let animationFrameRequest: number;
+	let start: number | null = null;
 
-			currentIndex++;
-			nextIndex = currentIndex + 1;
-			lastIndex = currentIndex - 1;
+	const play = () => {
+		if (!canAdvance()) return;
 
-			currentMemory = $memoryStore[currentIndex];
-			nextMemory = $memoryStore[nextIndex];
-			lastMemory = $memoryStore[lastIndex];
+		paused = false;
 
-			return true;
-		}
+		const draw = (now: number) => {
+			requestDraw();
 
-		return false;
-	};
+			start ??= now - progress * duration;
 
-	const toPreviousMemory = () => {
-		if (showPreviousMemory) {
-			resetAutoPlay();
+			const elapsed = now - start;
+			progress = Math.min(1, elapsed / duration);
 
-			currentIndex--;
-			nextIndex = currentIndex + 1;
-			lastIndex = currentIndex - 1;
+			if (progress !== 1) return;
 
-			currentMemory = $memoryStore[currentIndex];
-			nextMemory = $memoryStore[nextIndex];
-			lastMemory = $memoryStore[lastIndex];
-		}
+			start = now;
+
+			toNextAsset();
+
+			if (!canAdvance()) {
+				pause();
+				resetStart();
+			}
+		};
+
+		const requestDraw = () => (animationFrameRequest = requestAnimationFrame(draw));
```

Comment:
It is a little strange to define this function here, reference it above (before the definition), and also it's implementation references `draw` which also calls it. I think it'd be possible to actually have the entire `draw` method be defined outside of the `play` method, and `play` only starts the animationFrameRequest loop if necessary.

Reply:
Yeah, it is now. There was a point where I was working on it where `start` was defined in `play`. That's been moved out to handle pausing and navigation correctly, so I can now separate the function too.